rust/yarp-sys/tests/serialize_tests.rs

```diff
@@ -0,0 +1,102 @@
+use std::{ffi::CString, mem::MaybeUninit, path::Path};
+
+use yarp_sys::{
+    yp_buffer_free, yp_buffer_init, yp_buffer_t, yp_node_destroy, yp_parse, yp_parse_serialize,
+    yp_parser_free, yp_parser_init, yp_parser_t, yp_serialize,
+};
+
+fn ruby_file_contents() -> (CString, usize) {
+    let rust_path = Path::new(env!("CARGO_MANIFEST_DIR"));
+    let ruby_file_path = rust_path.join("../../lib/yarp.rb").canonicalize().unwrap();
+    let file_contents = std::fs::read_to_string(ruby_file_path).unwrap();
+    let len = file_contents.len();
+
+    (CString::new(file_contents).unwrap(), len)
+}
+
+#[test]
+fn serialize_test() {
+    let (ruby_file_contents, len) = ruby_file_contents();
+    let source = ruby_file_contents.as_ptr();
+    let mut parser = MaybeUninit::<yp_parser_t>::uninit();
+    let mut buffer = MaybeUninit::<yp_buffer_t>::uninit();
+
+    unsafe {
+        yp_parser_init(parser.as_mut_ptr(), source, len, std::ptr::null());
+        let parser = parser.assume_init_mut();
+        let node = yp_parse(parser);
+
+        if !yp_buffer_init(buffer.as_mut_ptr()) {
+            panic!("Failed to init buffer");
+        }
+
+        let buffer = buffer.assume_init_mut();
+        yp_serialize(parser, node, buffer);
+
+        let serialized = std::slice::from_raw_parts(buffer.value as *mut u8, buffer.length);
```

Comment:
with all of the unsafe code in the tests, I suggest running clippy on this crate. It likely will suggest to replace this `as` cast with `buffer.value.cast::<u8>()` among other things.

Reply:
this lint in particular that I've called out I believe requires warning on `clippy:pedantic`.

https://rust-lang.github.io/rust-clippy/master/index.html#/ptr_as_ptr

Reply:
Hm... I have my IDE setup to run clippy on every save and it doesn't tell me that. ...and if I run it from the CLI, I get no warnings.

Reply:
TIL... I added `#![warn(clippy::pedantic)]` (and a bunch of others) to `src/lib.rs` and ran `cargo clippy --tests`, and that came back with 0 warnings (after fixing other non-test things). I then ran `cargo clippy --tests -- -W "clippy:pedantic"` and got the warning you mentioned. All these years, I thought adding the `warn` attributes to my crate would apply to testing too.

Reply:
integration tests (the tests found in the `tests/` directory) are compiled as separate crates and thus must have their own crate-level attributes.